rfc/0019-hashed-mojmap.md

```diff
@@ -0,0 +1,97 @@
+## Motivation
+
+- One of the biggest, if not the biggest, reliance of Quilt on upstream Fabric is intermediary. For as long as Fabric keeps updating to new Minecraft versions, we will be behind them on updating because of this limitation. It is also not a good thing for a hard fork to have a reliance on upstream.
+
+- One of the primary goals of intermediary is to make sure that mods don't have to recompile between versions. However, when mods make the switch from Fabric to Quilt, they will have to recompile anyway. This gives us a unique opportunity to make breaking changes to the format of intermediary without breaking mods.
+
+- Intermediary was a good solution at the time of its inception. Since then, Mojang has provided us with their official mappings, with permission to use them as long as we don't redistribute them. Having to match intermediary between versions introduces a high maintenance cost for every snapshot and potentially also limits the bus factor of the whole project. The majority of the time updating Fabric to snapshots is typically spent matching.
+
+
+## Explanation
+
+Hashed Mojmap will produce base-62 hashes of information taken from Mojmap to produce mixed-case alphanumeric names like the following:
+
+- Classes: `Cls_hE89sr` (6 digits)
+- Fields: `fd_p9TXc40` (7 digits)
+- Methods: `md_fjH9ac8` (7 digits)
+
+To produce the hashes, the Mojmap names are processed as follows:
+
+1. Classes are converted from dotty format (`net.minecraft.a.Foo$Bar`) to slashy format (`net/minecraft/a/Foo$Bar`).
+1. Overriding methods in subclasses are discarded if present.
+1. Package names are removed from classes whose simple names are unique. That is, `net/minecraft/a/Bar` gets shortened to `Bar` if there is no other class named `Bar`, but if there are classes `net/minecraft/a/Foo` and `net/minecraft/b/Foo` that exist, then neither of them will be shortened. This includes references to these classes in method descriptors.
+1. Method descriptors are replaced with an empty string for methods whose name is unique within the class that the method is declared.
+1. Field names are converted to `fd_$x` where `$x` is the last 7 digits of the base-62 representation of the SHA-1 hash of the string `f$declaring_class.$field_name`.
+1. Method names are converted to `md_$x` where `$x` is the last 7 digits of the base-62 representation of the SHA-1 hash of the string `m$declaring_class.$method_name$descriptor`.
+1. Class names are converted to `Cls_$x` where `$x` is the last 6 digits of the base-62 representation of the SHA-1 hash of the class name.
+
+The above steps are designed to preserve some of the benefits of intermediary as much as possible without the need for matching. Testing has shown that around 50% of class renames in Mojmap are only repackages, which are unlikely to affect hashed Mojmap. The descriptor removal is designed to preserve method names where only the descriptor has changed, while ensuring hashed Mojmap names are all unique.
+
```

Comment:
Dot format would be considered a [binary name](https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.2.1) if we want to nitpick terms.

Reply:
Yes, but it's still somewhat ambiguous, as the class-internal representation of binary names uses slashes (same link).

Reply:
> Why slashy format instead of dotty format? I guess because we want to use JVM naming rather than Java-esque.

It's just to normalize it to one format or the other, to make sure everyone knows how to generate the same hash.

> Overriding: The hash would obviously have to be the same as the overwritten [sic] method.

Exactly, which is why overriding methods are discarded when computing the hashes, because they would produce a different hash because they have a different declaring class.

You are correct in saying that package names and method overloads will be broken once the number of simple names/method names goes from 1 to 2 or from 2 to 1. However I anticipate this being quite a rare event. It's certainly rare compared to package renames in general. Also for a hashed mojmap approach, it's this or not stripping prefixes at all, and stripping is surely the better option.

I will get back to you on package names once I have collected some data, but just from observation, most simple names are unique.